TESTING_PLAN.md (#32)
ggorlen/cs195-project: documentation/TESTING_PLAN.md

### Unit-level tests

- undo() should load the previous game state.

> - reset() should the current level to its original state.

Reviewer: ```suggestion
- reset() should load the current level to its original state.
```

> - move(U) should allow player to move up.

Reviewer: Should we specify here what is allowed for the player to move up. If a player moves up only when it's a goal or an empty cell. but if there's a wall, box-wall, box-box, etc, nothing should happen.

Reviewer 2: "previous game state" might not be entirely clear to a reader even if it's accurate -- maybe "undo the last move" makes more sense since that's what the user experiences. Game state is somewhat of an implementation detail.

> - move(U) should allow player to move up.

Reviewer: ```suggestion
- move(U) should allow the player to move up if it's a valid move.
```

- move(D) should allow player to move down.
- move(L) should allow player to move left.
- move(R) should allow player to move right.
- player should not be able to pass through impassable walls when move(U).
- player should not be able to pass through impassable walls when move(D).
- player should not be able to pass through impassable walls when move(L).
- player should not be able to pass through impassable walls when move(R).
- change_level(uint level_number) should allow user to switch levels.
- moves() should return the number of moves made on the current board.
- solved() should return true when all the boxes are on top of goal cells.

### Integration/E2E-level tests

- When level is solved() congratulatory message should appear.
- When level is solved() the user should be prompted to move to the next level or change to a different level.
- When pushed/clciked reset() should reset the board
- When pushed/clicked undo() should undo previous game state.

> - When pushed/clicked Start button should initiatie the deafult game by returnting level().

Reviewer: ```suggestion
- When pushed/clicked, Start button should initiate the default game by returning level().
```

> - When directional key up is pressed player should move up one.

Reviewer: ```suggestion
- When the directional key up is pressed player should move up one cell.
```

### Unit-level tests

> - Test that impassible walls are impassible by the player.

Reviewer: ```suggestion
- Test that impassable walls are impassable by the player.
```

> - Test that the level is complete/congratulatory message when and only when all boxes are moved to designated goal cells.

Reviewer 2: Congratulatory message is purely a JS/UI concept, so the game engine only needs to return true or false when the level is complete via the `solved()` func. Feel free to move this to the front-end integration tests though.

> - Test when a level is complete, the user is prompted to move to the next level or jump to a different level.

Reviewer 2: Same as above -- prompts are UI concepts that the game engine doesn't care about.

- Test that all boxes are pushable by the player.
- Test the score is tracking player movement.

> - Test that the undo button undoes moves.

Reviewer: ```suggestion
- Test that the undo() loads the previous game state.
```

> - Test the reset button reset’s the board properly.

Reviewer 2: Same as above, the backend doesn't understand what a button is, so we can move this to integration testing.

> - Test the reset button reset’s the board properly.

Reviewer: ```suggestion
- Test that the reset() loads the initial game state of the specified level.
```

- Test that the 'Sart' button initiates the default game by dropping the player into level 1.
- Test that the drop-down menu is clickable by the player and allows them to jump to different levels.

> - Test that the directional keys move the character in the proper direction up, down, left, and right.

Reviewer: ```suggestion
- Test that the move() handles the logic of moving the character in the proper direction up, down, left, and right.
```

### Unit-level tests

- undo() should load the previous game state.
- reset() should load the current level to its original state.
- move(U) should allow the player to move up if it's a valid move.
- move(D) should allow player to move down.
- move(L) should allow player to move left.
- move(R) should allow player to move right.
- player should not be able to pass through impassable walls when move(U).
- player should not be able to pass through impassable walls when move(D).
- player should not be able to pass through impassable walls when move(L).
- player should not be able to pass through impassable walls when move(R).
- change_level(uint level_number) should allow user to switch levels.
- moves() should return the number of moves made on the current board.

> - solved() should return true when all the boxes are on top of goal cells.

Reviewer 2: Maybe we should throw in some box pushing tests -- the 4 directions for each push are good, but also somewhat redundant as far as the requirements for 10 tests go.

Good function names here, but I'd probably keep the function call/code out of it so it's a bit more high-level -- maybe "should undo a move", "should disallow moving into a wall from the right", "should detect a solved board", "should detect an unsolved board", etc.

### Integration/E2E-level tests

> - When level is solved() congratulatory message should appear.

Reviewer 2: Here, we aren't testing the back-end functions like `solved()` -- our test suite is pretending to be the user that isn't aware of technical details, just sees squares and UI. So, "should show congratulatory message when level is solved", "should display next level after clicking congraulatory message", "should move a player to the left when left arrow is pressed and destination square is empty", etc.

- When level is solved() the user should be prompted to move to the next level or change to a different level.

> - When pushed/clciked reset() should reset the board

Reviewer 2: ```suggestion
- should reset the board by pressing the reset button
```

- Should allow the player to push a box right onto an empty goal cell.
- Should detect a solved board.
- Should detect an unsolved board.
- Should allow the user to switch levels.
- Should return the number of moves made on the current board.

### Integration/E2E-level tests

- Should show congratulatory message when a level is solved.
- Should allow the player to click on the desired level to play.
- Should display next level after clicking congratulatory message or change to a different level.
- Should reset the board by pressing the reset button.
- Should undo previous game state when pressing the undo button.
- Should initiate the default game when pressing the start button.
- When the directional key up is pressed, the player should move up one cell.
- Should move the player left when the left arrow is pressed and the destination square is empty.
- Should move the player right when the right arrow is pressed and the destination square is empty.
- Should move the player up when the up arrow is pressed and the destination square is empty.
- Should move the player down when the down arrow is pressed and the destination square is empty.

> - Should allow the player an obvious way to restart or exit the game besides closing the window.

Author: ```suggestion
- Should allow the user to exit from a level to the start menu.
```
